Package manager backend: detect package dll lock

huguesv opened the pull request:
I haven't gotten the UI part to work successfully in the case where the user wants us to reset the repl for them before install/uninstall (Task returned BeginInteractionAsync is awaited forever), so it's just the backend + tests in this PR.

FYI RestartManager impl comes from VS.

int19h commented on src/Host/Client/Impl/Host/RHost.cs:
```diff
@@ -50,6 +50,8 @@ public sealed partial class RHost : IDisposable, IRExpressionEvaluator {
         private TaskCompletionSource<object> _cancelAllTcs;
         private CancellationTokenSource _cancelAllCts = new CancellationTokenSource();
 
+        public int ProcessId { get { return _process != null ? _process.Id : 0; } }
```
You can use `=>` here to make the one-liner slightly shorter:
```cs
public int ProcessId => ...
```

int19h replied:
Also, perhaps it's better to make it `int?`, and return `null` if there's no host.

int19h replied:
Or maybe even just expose the process directly... @AlexanderSher, I recall we talked about something where having the process accessible could be advantageous. Should we just do that?

int19h commented on src/R/Components/Impl/PackageManager/Implementation/RestartManager.cs:
```diff
@@ -0,0 +1,143 @@
+using System.Security;
+
+namespace Microsoft.R.Components.PackageManager.Implementation {
+    internal class RestartManager {
```
Looks like this could be `static`.

int19h replied:
Also, this is very generic, and would probably be a better fit in `Microsoft.Common.Core`.

int19h commented on src/R/Components/Impl/PackageManager/Implementation/RestartManager.cs:
```diff
@@ -0,0 +1,143 @@
+            uint sessionHandle;
+            List<Process> processes = new List<Process>();
+
+            int restartManager = NativeMethods.RmStartSession(out sessionHandle, 0, Guid.NewGuid().ToString("N"));
```
This is really a status/error code, right? Can you rename it to `error` or something similarly descriptive?

int19h commented on src/R/Components/Impl/PackageManager/Implementation/RestartManager.cs:
```diff
@@ -0,0 +1,143 @@
+                uint pnProcInfoNeeded = 0, pnProcInfo = 0, lpdwRebootReasons = RmRebootReasonNone;
+                restartManager = NativeMethods.RmGetList(sessionHandle, out pnProcInfoNeeded, ref pnProcInfo, null, ref lpdwRebootReasons);
+
+                if (restartManager == ERROR_MORE_DATA) {
```
This likely needs to be a loop, if the list can change between the calls. Basically keep trying until it fits.

int19h commented on src/R/Components/Impl/PackageManager/Implementation/RestartManager.cs:
```diff
@@ -0,0 +1,143 @@
+                    RM_PROCESS_INFO[] processInfo = new RM_PROCESS_INFO[pnProcInfoNeeded];
+                    pnProcInfo = (uint)processInfo.Length;
+
+                    restartManager = NativeMethods.RmGetList(sessionHandle, out pnProcInfoNeeded, ref pnProcInfo, processInfo, ref lpdwRebootReasons);
```
Again, assuming the list can change between consecutive `RmGetList` calls, this might end up filling fewer than `pnProcInfoNeeded` items in the array - I'm assuming that `pnProcInfo` is `ref` precisely so that the function can communicate the actual number of items it has filled. The loop below should take that into account.

int19h commented on src/R/Components/Impl/PackageManager/Implementation/RestartManager.cs:
```diff
@@ -0,0 +1,143 @@
+        /// </summary>
+        /// <param name="filePaths">Filenames with absolute path.</param>
+        /// <returns>List of processes locking the files. null if it encounters any error.</returns>
+        public static IList<Process> GetProcessesUsingFiles(string[] filePaths) {
```
I would make this `IEnumerable<Process>`, and use `yield return` inside to fetch them on demand. The caller can then do `ToList` or not, as needed.

huguesv commented:
I think I've addressed all feedback.  Kept the process id as (nullable) int rather than returning Process object.  My thinking is that if we don't need it right now, we'll avoid potential misuse (someone accessing stuff through the Process rather than going through our existing host APIs).

int19h commented:
:+1:
